### PR TITLE
Greatly improve System.arraycopy + Arrays.fill in haxe-cpp target + some more improvements on pure-cpp target

### PR DESCRIPTION
**Pull request by soywiz**
```
WIN-JAVA 1.8:

arraycopy byte...8.964731991291046
arraycopy short...25.391018986701965
arraycopy char...25.71083700656891
arraycopy int...28.533034026622772
arraycopy float...27.600202977657318

/////////////////////////////////

WIN-JTRANSC-CPP 0.6.4:

arraycopy byte...1.0
arraycopy short...3.0
arraycopy char...3.0
arraycopy int...9.0
arraycopy float...10.0

/////////////////////////////////

WIN-JTRANSC-HAXE-CPP 0.6.4:

arraycopy byte...107.00048828125
arraycopy short...87.999267578125
arraycopy char...85.0009765625
arraycopy int...10.000244140625
arraycopy float...89.000732421875

/////////////////////////////////

WIN-JTRANSC-HAXE-CPP 0.6.4 (with this PR):

arraycopy byte...3.000244140625
arraycopy short...4.9990234375
arraycopy char...5.001220703125
arraycopy int...10.999755859375
arraycopy float...10.99951171875
arraycopy double...20.000244140625
arraycopy long...17.999755859375
```

### CODE REVIEW

**Comment by soywiz** [comment]
I just enabled Haxe-cpp travis tests. It seems that fail, and seems that failed before too (with big endian double reading), so I have to check it out first.

**Comment by soywiz** [comment]
Required for: https://github.com/jtransc/gdx-backend-jtransc/issues/59

**Comment by soywiz** [comment]
A way for improving this even further would be to add `JA_0.pointer` and `JA_0.size`, and on System.arraycopy just use that information instead of doing dynamic dispatch. Also I would avoid extra checkings (that I have already done) from the `blit` function.

**Comment by soywiz** [comment]
@SergeyLabutin This should be ready. Check it and if it looks good to you, merge it (please, select **Squash and merge** instead of ~~create merge commit~~ to just add one combined commit). After these changes. Copying and filling should be much faster now for all types in both pure C++ and Haxe-C++. I have slightly improved overall performance on other areas like Object array access that should be now faster in Haxe-Cpp:

WIN-JAVA 8:
```
JTransc 0.6.4 - java
Java 1.8.0_121 - 25.121-b13 - 1.8.0_121-b13
freeMemory: 254741016, maxMemory: 3791650816, totalMemory: 257425408
Benchmarking:
plain loops...3.1587889790534973
shift left constant...6.1596759557724
shift right constant...6.1155500411987305
shift unsigned right constant...6.309554994106293
shift left constant long...6.454694986343384
shift right constant long...6.394888997077942
shift unsigned right constant long...6.148736000061035
left shift...13.430141985416412
right shift...13.392944991588593
right unsigned shift...13.7605339884758
call static mult...18.026467978954315
call instance mult...18.66209101676941
call instance div...100.06819301843643
instanceof classes...14.770312011241913
arraycopy byte...9.158374011516571
arraycopy short...26.48175400495529
arraycopy char...26.838768005371094
arraycopy int...30.459960997104645
arraycopy float...29.908577024936676
arraycopy double...41.25534701347351
arraycopy long...42.622502982616425
arraycopy object...43.64504301548004
arrayfill byte...1.4597820043563843
arrayfill short...2.669034004211426
arrayfill char...2.699666976928711
arrayfill int...7.529748976230621
arrayfill float...7.348506987094879
arrayfill double...12.290906965732574
arrayfill long...12.147953987121582
arrayfill object...57.01428401470184
write byte[]...3.613898992538452
write short[]...3.639792025089264
write char[]...3.69303297996521
write int[]...3.710173010826111
write float[]...4.806011974811554
write double[]...6.006148993968964
String Builder 1...22.482394993305206
String Builder 2...6.8423410058021545
long arithmetic...1.3175600171089172
simd mutable...11.907635033130646
simd immutable...69.06013000011444
simd mutable matrix mult...57.659754037857056
StringBuilder1...20.065349996089935
StringBuilder2...17.18079298734665
Non Direct Buffer...3.0785619616508484
Direct Buffer Int/float...1.440818965435028
Direct Buffer Short/Char...1.590699017047882
Direct Buffer Double/Long...1.8434170484542847
FastMemory...1.0947459936141968
Create Instances1 local...0.0025529861450195312
Create Instances2 local...0.003282010555267334
Create Instances2 global...4.36585396528244
Create Instances with builder...15.424533009529114
Java's CRC32...6.448495984077454
jzlib's CRC32...60.84662199020386
compress java's Deflate...16.958343982696533
compress jzlib...22.043330013751984
random...7.129338026046753
exception...6.53236997127533
TOTAL time: 967.170766890049
```

WIN-PURE-CPP:
```
JTransc 0.6.4 - cpp
Java 1.8.0_51 - 25.51-b03 - 1.8.0_51-b16
freeMemory: 69632, maxMemory: 156424, totalMemory: 156424
Benchmarking:
plain loops...2.0
shift left constant...3.0
shift right constant...6.0
shift unsigned right constant...7.0
shift left constant long...12.0
shift right constant long...15.0
shift unsigned right constant long...16.0
left shift...24.0
right shift...25.0
right unsigned shift...23.0
call static mult...16.0
call instance mult...25.0
call instance div...104.0
instanceof classes...91.0
arraycopy byte...2.0
arraycopy short...3.0
arraycopy char...3.0
arraycopy int...9.0
arraycopy float...10.0
arraycopy double...18.0
arraycopy long...19.0
arraycopy object...9.0
arrayfill byte...2.0
arrayfill short...3.0
arrayfill char...4.0
arrayfill int...7.0
arrayfill float...6.0
arrayfill double...64.0
arrayfill long...64.0
arrayfill object...6.0
write byte[]...6.0
write short[]...5.0
write char[]...5.0
write int[]...1.0
write float[]...3.0
write double[]...6.0
String Builder 1...35.0
String Builder 2...18.0
long arithmetic...2.0
simd mutable...14.0
simd immutable...12.0
simd mutable matrix mult...24.0
StringBuilder1...97.0
StringBuilder2...53.0
Non Direct Buffer...5.0
Direct Buffer Int/float...4.0
Direct Buffer Short/Char...4.0
Direct Buffer Double/Long...4.0
FastMemory...4.0
Create Instances1 local...41.0
Create Instances2 local...37.0
Create Instances2 global...43.0
Create Instances with builder...527.0
Java's CRC32...61.0
jzlib's CRC32...60.0
compress java's Deflate...23.0
compress jzlib...25.0
random...5.0
exception...48.0
TOTAL time: 1770.0
```

WIN-HAXE-CPP:
```
JTransc 0.6.4 - haxe
Java 1.8.0_51 - 25.51-b03 - 1.8.0_51-b16
freeMemory: 8589934592, maxMemory: 8589934592, totalMemory: 8589934592
Benchmarking:
plain loops...10.000244140625
shift left constant...12.001220703125
shift right constant...12.0009765625
shift unsigned right constant...10.000244140625
shift left constant long...37.000732421875
shift right constant long...34.00048828125
shift unsigned right constant long...32.9990234375
left shift...18.999267578125
right shift...17.99951171875
right unsigned shift...18.9990234375
call static mult...21.99951171875
call instance mult...13.999755859375
call instance div...91.9990234375
instanceof classes...1101.99926757813
arraycopy byte...2.000732421875
arraycopy short...3.999755859375
arraycopy char...4.9990234375
arraycopy int...10.000244140625
arraycopy float...8.9990234375
arraycopy double...18.9990234375
arraycopy long...18.999267578125
arraycopy object...17.999755859375
arrayfill byte...1.9990234375
arrayfill short...3.999755859375
arrayfill char...3.999755859375
arrayfill int...6.999755859375
arrayfill float...6.00048828125
arrayfill double...35.999267578125
arrayfill long...17.99951171875
arrayfill object...62.00048828125
write byte[]...12.0009765625
write short[]...12.0009765625
write char[]...13.00048828125
write int[]...11.999267578125
write float[]...11.9990234375
write double[]...13.00048828125
String Builder 1...272.00146484375
String Builder 2...67.9990234375
long arithmetic...2.000732421875
simd mutable...35.0
simd immutable...422.998291015625
simd mutable matrix mult...114.000244140625
StringBuilder1...300.001220703125
StringBuilder2...298.00048828125
Non Direct Buffer...10.000244140625
Direct Buffer Int/float...9.0009765625
Direct Buffer Short/Char...9.000732421875
Direct Buffer Double/Long...34.00048828125
FastMemory...7.999267578125
Create Instances1 local...42.999267578125
Create Instances2 local...44.000732421875
Create Instances2 global...52.99951171875
Create Instances with builder...471.000732421875
Java's CRC32...62.000244140625
jzlib's CRC32...64.99853515625
compress java's Deflate...39.9990234375
compress jzlib...41.000244140625
random...66.0
exception...20.00048828125
TOTAL time: 4219.99633789063
```

File: jtransc-gen-haxe/src/com/jtransc/gen/haxe/HaxeTarget.kt [hunk 1]
```diff
@@ -815,7 +818,7 @@ class HaxeGenerator(injector: Injector) : CommonGenerator(injector) {
 	override val IntType = "Int"
 	override val FloatType = "Float32"
 	override val DoubleType = "Float64"
-	override val LongType = "haxe.Int64"
+	override val LongType = "NativeInt64"
```
Comment (intrigus) [inline jtransc-gen-haxe/src/com/jtransc/gen/haxe/HaxeTarget.kt]: What's the difference between those?

File: jtransc-main/test/javatest/utils/FillTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	final float[] arrayFloat = new float[size];
+	final double[] arrayDouble = new double[size];
+	final Object[] arrayObject = new Object[size];
+	final Demo[] demos = new Demo[size];
+
+	public FillTest() {
+		for (int n = 0; n < size; n++) demos[n] = new Demo(n);
+		//System.out.println(demos instanceof Object[]);
+		//System.out.println(demos instanceof Demo[]);
+		//System.out.println(demos instanceof Runnable[]); // This fails!
```
Comment (intrigus) [inline jtransc-main/test/javatest/utils/FillTest.java]: Does this still fail?
Could you create an issue for this?

File: jtransc-rt/resources/hx/N.hx [hunk 3]
```diff
@@ -27,31 +26,95 @@ typedef JavaDouble = {% CLASS java.lang.Double %}
...
+	inline static int _i2c(int v) { return (unsigned short)v; };
+	inline static void _memcpy(unsigned char *dst, int dstPos, unsigned char *src, int srcPos, int shift, int count) {
+		::memcpy(
+			&dst[dstPos << shift],
+			&src[srcPos << shift],
+			count << shift
+		);
+	};
+
+	inline static void _memset_N2(unsigned char *dst, int dstPos, int count, int value) {
```
Comment (intrigus) [inline jtransc-rt/resources/hx/N.hx]: Can you not use intNN_t for the types?
Or is this haxe code?

File: jtransc-rt/resources/hx/N.hx [hunk 4]
```diff
@@ -27,31 +26,95 @@ typedef JavaDouble = {% CLASS java.lang.Double %}
...
+		float *ptr = &(((float *)dst)[dstPos]);
+		for (int n = 0; n < count; n++) ptr[n] = (float)value;
+	};
+
+	inline static long long int _ldivRaw(long long int a, long long int b) {
+		return a / b;
+	};
+
+	inline static long long int _lnew(int high, int low) {
+		//printf("long long int: %d\\n", sizeof(long long int));
```
Comment (intrigus) [inline jtransc-rt/resources/hx/N.hx]: Remove this comment or hide it with a debug flag?

File: jtransc-rt/resources/hx/N.hx (hunk 3 above)
Comment (soywiz) [inline jtransc-rt/resources/hx/N.hx]: C++ generated from Haxe. I have tried `intNN_t` and they are not defined. Which is a bit, meh. Not sure that `long long int` resolves to `int64_t` in all targets.

File: jtransc-rt/resources/hx/N.hx (hunk 4 above)
Comment (soywiz) [inline jtransc-rt/resources/hx/N.hx]: This is not required anymore. So I'm going to delete this.

File: jtransc-gen-haxe/src/com/jtransc/gen/haxe/HaxeTarget.kt (hunk 1 above)
Comment (soywiz) [inline jtransc-gen-haxe/src/com/jtransc/gen/haxe/HaxeTarget.kt]: This is not required anymore. So I'm going to delete this.

File: jtransc-main/test/javatest/utils/FillTest.java (hunk 2 above)
Comment (soywiz) [inline jtransc-main/test/javatest/utils/FillTest.java]: https://github.com/jtransc/jtransc/issues/192

**Comment by soywiz** [comment]
@intrigus I think I have fixed all your concerns

**Comment by SergeyLabutin** [comment]
Good, i test tommorow, again today not completed, but only 0-0,45% boost, i think without copy will better.

**Comment by soywiz** [comment]
If it doesn't improve, then maybe bottleneck is elsewhere. But at any rate, this won't hurt. So even if not improves much in your case, I think it is worth the merge unless it breaks something.

**Pull request merged**